src/frontend/screens/Settings/ProjectSettings/DeviceName/index.tsx

```diff
@@ -0,0 +1,223 @@
+  },
+  discardChangesAlertConfirmText: {
+    id: 'screens.Setting.ProjectSettings.DeviceName.discardChangesAlertConfirmText',
+    defaultMessage: 'Discard Changes',
+  },
+  discardChangesAlertCancelText: {
+    id: 'screens.Setting.ProjectSettings.DeviceName.discardChangesAlertCancelText',
+    defaultMessage: 'Continue Editing',
+  },
+});
+
+export type NavigationParams = {mode: 'display' | 'edit'};
+
+export const DeviceName: NativeNavigationComponent<'DeviceName'> = ({
+  route,
+  navigation,
+}) => {
+  const {mode} = route.params;
+  const {data} = useDeviceInfo();
+
+  return (
+    <ScrollView contentContainerStyle={styles.container}>
+      {data?.name &&
+        (mode === 'display' ? (
+          <DisplayMode deviceName={data.name} navigation={navigation} />
+        ) : (
+          <EditMode deviceName={data.name} navigation={navigation} />
+        ))}
+    </ScrollView>
+  );
+};
+
+function DisplayMode({
+  deviceName,
+  navigation,
+}: {
+  deviceName: string;
+  navigation: React.ComponentProps<
+    NativeNavigationComponent<'DeviceName'>
+  >['navigation'];
+}) {
+  const {formatMessage: t} = useIntl();
+
+  React.useEffect(
+    function updateNavigationOptions() {
+      navigation.setOptions({
+        headerRight: () => (
+          <IconButton onPress={() => navigation.setParams({mode: 'edit'})}>
+            <EditIcon />
+          </IconButton>
+        ),
+      });
+    },
+    [navigation],
+  );
+
+  return (
+    <FieldRow label={t(m.deviceNameDisplayHeader)}>
+      <Text>{deviceName}</Text>
+    </FieldRow>
+  );
+}
+
+DeviceName.navTitle = m.title;
+
+function EditMode({
+  deviceName,
+  navigation,
+}: {
+  deviceName: string;
+  navigation: React.ComponentProps<
+    NativeNavigationComponent<'DeviceName'>
+  >['navigation'];
+}) {
+  const {formatMessage: t} = useIntl();
+  const {mutate: updateDeviceName} = useEditDeviceInfo();
+
+  const {control, getValues, handleSubmit} = useForm<{
+    deviceName: string;
+  }>({
+    defaultValues: {deviceName},
+  });
+
+  React.useEffect(
+    function showDiscardChangesAlert() {
+      const unsubscribe = navigation.addListener('beforeRemove', event => {
+        event.preventDefault();
+
+        const nameHasChanges = getValues('deviceName') !== deviceName;
+
+        if (nameHasChanges) {
+          Alert.alert(t(m.discardChangesAlertTitle), undefined, [
+            {style: 'cancel', text: t(m.discardChangesAlertCancelText)},
+            {
+              style: 'default',
+              text: t(m.discardChangesAlertConfirmText),
+              onPress: () => {
+                // TODO: Is going to display mode desired behavior here? Or should it go back to previous screen
+                navigation.setParams({mode: 'display'});
+              },
+            },
+          ]);
+        } else {
+          // TODO: Is going to display mode desired behavior here? Or should it go back to previous screen
+          navigation.setParams({mode: 'display'});
```

Comment:
highlighting these as it's related to the navigation behavior when exiting the "editing" mode either through (1) cancelling your changes or (2) saving your changes. This implementation currently assumes that we want to switch to the "display" mode in both cases, as opposed to navigating to some other screen (or the previous screen)